File: amp.py
import io
import logging
import threading
from dataclasses import dataclass
from enum import Enum
from typing import ClassVar, Union, Optional, Any, Dict, List, Tuple, Iterator, Type

import serial


class SerialBaudRate(Enum):
    B_9600 = 9600
    B_19200 = 19200
    B_38400 = 38400
    B_57600 = 57600
    B_115200 = 115200
    B_230400 = 230400


class SerialBaudOptions(Enum):
    AUTO = 'auto'


class SerialBaudAdjustOptions(Enum):
    OFF = 'off'
    MAX = 'max'


@dataclass(frozen=True)
class Range:
    min: int
    max: int

    def __contains__(self, i: int):
        return self.min <= i <= self.max

    def __str__(self):
        return f'[{self.min},{self.max}]'

    def __iter__(self) -> Iterator[int]:
        return iter(range(self.min, self.max+1))


@dataclass(init=False, frozen=True)
class ZoneId:
    amp: int
    zone: int

    VALID_AMPS: ClassVar[Range] = Range(1, 3)
    VALID_ZONES: ClassVar[Range] = Range(1, 6)

    def __init__(self, zone_id: Union[int, str, 'ZoneId']) -> None:
        zone_id = int(zone_id)
        object.__setattr__(self, 'amp', int(zone_id / 10))
        object.__setattr__(self, 'zone', zone_id % 10)

        if self.amp not in self.VALID_AMPS:
            raise ValueError(f'Amp {self.amp} is not within accepted range {self.VALID_AMPS}.')

        if self.zone not in self.VALID_ZONES:
            raise ValueError(f'Amp {self.amp} zone {self.zone} is not within accepted range {self.VALID_ZONES}.')

    def topic_fragment(self):
        return f'amps/{self.amp}/zones/{self.zone}'

    def __str__(self):
        return f'{self.amp}{self.zone}'

    def __int__(self):
        return (self.amp * 10) + self.zone


class AmpId(ZoneId):
    VALID_ZONES = Range(0, 0)

    @classmethod
    def for_zone(cls, zone: ZoneId):
        return cls(zone.amp * 10)

    def topic_fragment(self):
        return f'amps/{self.amp}'

    def zones(self) -> List[ZoneId]:
        return [ZoneId((self.amp * 10) + z) for z in ZoneId.VALID_ZONES]


@dataclass(frozen=True)
class ZoneAttribute:
    name: str
    type: Union[Type[int], Type[bool]]
    key: Optional[str] = None
    offset: int = 0
    range: Range = Range(0, 99)

    def _check_value(self, value: Any):
        if not isinstance(value, self.type):
            raise ValueError(f'{self.name}: expected value of type {self.type}. Got a {type(value)}.')

        if self.type == int:
            if value not in self.range:
                raise ValueError(f'{self.name}: {value} is not within accepted range {self.range}.')

    def decode_value(self, data: str) -> Any:
        value = self.type(int(data, base=10) + self.offset)
        self._check_value(value)

        return value

    def encode_value(self, value: Union[int, bool]) -> str:
        self._check_value(value)

        value = int(value) - self.offset

        return f'{value:0=2d}'


# order matters. listed in order of zone enquiry response
ZONE_ATTRIBUTES = [
    ZoneAttribute('public_announcement', bool),
    ZoneAttribute('power', bool, key='PR'),
    ZoneAttribute('mute', bool, key='MU'),
    ZoneAttribute('do_not_disturb', bool, key='DT'),
    ZoneAttribute('volume', int, key='VO', range=Range(0, 38)),
    ZoneAttribute('treble', int, key='TR', offset=-7, range=Range(-7, 7)),
    ZoneAttribute('bass', int, key='BS', offset=-7, range=Range(-7, 7)),
    ZoneAttribute('balance', int, key='BL', offset=-10, range=Range(-10, 10)),
    ZoneAttribute('source', int, key='CH', range=Range(1, 6)),
    ZoneAttribute('keypad_connected', bool)
]
# only attributes with keys are settable
SETTABLE_ZONE_ATTRIBUTES = list(filter(lambda a: a.key is not None, ZONE_ATTRIBUTES))

ZoneEnquiryResponseType = Dict[ZoneId, Dict[ZoneAttribute, Any]]


class MwhaAmpConnection:
    CHUNK_SEP = b'\r\n#'
    CHUNK_SEP_LEN = len(CHUNK_SEP)

    lock = threading.Lock()

    def __init__(self, port: str,
                 baud: Union[SerialBaudRate, SerialBaudOptions] = SerialBaudOptions.AUTO,
                 adjust_baud: Union[SerialBaudRate, SerialBaudAdjustOptions] = SerialBaudAdjustOptions.MAX,
                 reset_baud: bool = True,
                 read_timeout: float = 1.0):

        self.serial = serial.serial_for_url(port,
                                            baudrate=9600 if baud == SerialBaudOptions.AUTO else baud.value,
                                            bytesize=serial.EIGHTBITS,
                                            parity=serial.PARITY_NONE, stopbits=serial.STOPBITS_ONE,
                                            xonxoff=False, rtscts=False, dsrdtr=False,
                                            timeout=read_timeout)

        self.logger = logging.getLogger(MwhaAmpConnection.__name__)

        # detect baud
        self.previous_baud = self.detect_baud() if SerialBaudOptions.AUTO else baud

        # adjust baud, if requested
        if adjust_baud == SerialBaudAdjustOptions.MAX:
            self.set_baud(list(SerialBaudRate)[-1])

        elif adjust_baud == SerialBaudAdjustOptions.OFF:
            pass

        else:
            self.set_baud(adjust_baud)

        self.reset_baud = reset_baud

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    def close(self):
        if self.reset_baud:
            self.set_baud(self.previous_baud)

        self.serial.close()

    def exec_command(self, cmd: str, expected_responses: int) -> List[str]:
        with self.lock:
            self.serial.reset_input_buffer()
            self.serial.reset_output_buffer()

            raw_cmd = (cmd + '\r').encode('ascii')
            self.serial.write(raw_cmd)
            self.serial.flush()

            def read_response():
                # serial.read_until will timeout if the expected terminal string isn't received
                # within the timeout period, requiring a timeout that is big enough to receive all
                # the data (unknown).
                # Instead, read is called directly to fetch a single byte.
                # If data is still being received but slowly it's not a timeout.
                # It's only a problem if the single byte read times out.

                buf = b''
                while True:
                    ch = self.serial.read(1)
                    if len(ch) == 0:
                        raise serial.SerialTimeoutException(
                            f'Timeout occurred while reading response for command "{cmd}".')

                    buf += ch
                    if buf[-3:] == b'\r\n#':
                        break

                buf = buf[:-3]

                if buf == b'\r\nCommand Error.':
                    raise IOError(f'Command error occurred while executing command "{cmd}".')

                return buf.decode('ascii')

            echo = read_response()
            if echo != cmd:
                raise IOError(f'Serial echoback was not the expected value. Expected "{cmd}" got "{echo}".')

            return [read_response() for _ in range(expected_responses)]

    def set_baud(self, rate: SerialBaudRate):
        self.logger.info('Setting baud rate to %s', rate.value)

        self.serial.reset_input_buffer()
        self.serial.reset_output_buffer()

        self.serial.write(f'<{rate.value}\r'.encode('ascii'))
        self.serial.flush()

        # As soon as the amp receives the '\r' of the command it switches baud.
        # There's no way to sync switching local baud with the amp, to my knowledge, esp. over IP.
        # Hence, even though baud set commands return "#Done." on success, the response is almost always corrupted.
        # Instead, drain the input buffer and resync after changing baud...
        while len(self.serial.read(1)) > 0:
            pass

        self.serial.baudrate = rate.value

        self._resync()

    def _resync(self):
        """
        Re-synchronise the serial stream. Send up to 4 empty commands (equiv of just '\r') and eventually expect an
        empty, yet valid reply.
        """
        for i in range(4):
            try:
                self.exec_command('', 0)
                return

            except IOError as e:
                if i == 3:
                    raise IOError('Unable to resync serial port.') from e

                continue

    def detect_baud(self) -> SerialBaudRate:
        """
        Detect the current baud rate of the amp.
        This is always 9600 after power on but an abruptly closed serial session could leave the baud set to a higher
        rate.

        Since the amp echos back what is written on the serial line, detection is done by writing a known string and
        trying to read the echo back. If the response is corrupted, try a different baud rate.
        """
        test_data = b'baudrate detect'

        detected_rate = None

        for rate in list(SerialBaudRate):
            self.logger.info('Trying baud rate %s', rate.value)

            self.serial.baudrate = rate.value

            self.serial.write(test_data)
            echo = self.serial.read(len(test_data))
            if echo == test_data:
                detected_rate = rate
                break

            self.logger.debug('%s =! %s', test_data, echo)

        if detected_rate is None:
            raise IOError('Unable to detect current baud rate.')

        self.logger.info('Detected current baud rate as %s', detected_rate.value)

        self._resync()

        return detected_rate

    def zone_enquiry(self, zone_id: Union[ZoneId, AmpId]) -> ZoneEnquiryResponseType:
        """
        Read the status of a zone, or all zones on an amp.

        When passed a ZoneID, will return the status of that zone.
        When passed an AmpID, will return the status of all zones on that amp.
        """
        expected_responses = 6 if isinstance(zone_id, AmpId) else 1

        def decode(data: str) -> Tuple[ZoneId, Dict[ZoneAttribute, Any]]:
            """Parse response: `>xxaabbccddeeffgghhiijj`"""
            s = io.StringIO(data)
            s.read(1)  # prefix '>'

            zid = ZoneId(s.read(2))
            attributes = {a: a.decode_value(s.read(2)) for a in ZONE_ATTRIBUTES}

            return zid, attributes

        zone_attributes = [decode(r) for r in self.exec_command(f'?{zone_id}', expected_responses)]

        return dict(zone_attributes)

    def zone_set(self, zone_id: ZoneId, attribute: ZoneAttribute, value: Any) -> None:
        """
        Set the value of a zone attribute.

        When passed a ZoneID, will set the attribute on that zone.
        When passed an AmpID, will set the attribute on all zones on that amp.
        """
        if attribute.key is None:
            raise ValueError(f"Can't set attribute {zone_id}{attribute.name}.")

        encoded = attribute.encode_value(value)

        self.exec_command(f'<{zone_id}{attribute.key}{encoded}', 0)

File: mwha2mqttd/src/amp.rs

use std::ascii::escape_default;
use std::io::Read;
use std::io::Write;

use std::net::TcpStream;
use std::str;

use anyhow::bail;
use itertools::Itertools;
use log::debug;

use anyhow::{Context, Result};

use common::zone::ZoneId;
use common::zone::ZoneAttribute;



pub trait Port: Read + Write + Send {}

impl Port for TcpStream {}


pub struct ZoneStatus {
    pub zone_id: ZoneId,
    pub attributes: Vec<ZoneAttribute>
}


pub struct Amp {
	port: Box<dyn Port>
}

fn escape(s: &String) -> String {
    String::from_utf8(
        s.bytes()
            .flat_map(|b| std::ascii::escape_default(b))
            .collect::<Vec<u8>>(),
    )
    .unwrap()
}

pub fn print_buffer(buffer: &[u8]) {
    let foo = &buffer.iter()
            .flat_map(|b| escape_default(*b))
            .collect::<Vec<u8>>();

        let s = String::from_utf8_lossy(
            &foo
        );
        print!("{}, {:?}", s, buffer);
}

impl Amp {
    const END_OF_RESPONSE_MARKER: &[u8] = b"\r\n#";

	pub fn new(port: Box<dyn Port>) -> Result<Self> {
        let mut amp = Self {
			port
		};

        amp.resync().context("failed to resync amp connection")?;

		Ok( amp )
	}

    fn read_until(&mut self, marker: &[u8]) -> Result<Vec<u8>> {
        let mut buffer = Vec::with_capacity(256);
		
        // maybe switch to a BufReader?
        // (but this is 9600 baud serial, performance isn't really an issue!)
        while !buffer.ends_with(marker) {
            let mut ch = [0; 1];

            self.port.read(&mut ch)
                .context("failed to read from port")?;
            
            buffer.extend_from_slice(&ch);
        }

        Ok(buffer)
    }

    fn read_command_response(&mut self) -> Result<Vec<u8>> {
        let mut buffer = self.read_until(Self::END_OF_RESPONSE_MARKER)?;

        buffer.truncate(buffer.len() - Self::END_OF_RESPONSE_MARKER.len());

        if buffer == b"\r\nCommand Error." {
            bail!("amp responded with command error while executing command.");
        }

        Ok(buffer)
    }

	fn exec_command(&mut self, command: &[u8], expected_responses: usize) -> Result<Vec<Vec<u8>>> {
		// write command
        self.port.write(command)?;
		self.port.write(b"\r")?;
		self.port.flush()?;
		
        // read echoback
		let echo = self.read_command_response()?;
        if echo != command {
            bail!("serial echoback was not the expected value. got = {:?}, expected = {:?}", str::from_utf8(&echo), str::from_utf8(command));
        }

        // read responses
        let mut responses = Vec::with_capacity(expected_responses);
        for _i in 0..expected_responses {
            responses.push(self.read_command_response()?);
        }

		Ok(responses)
	}

    /// Resyncronise the serial stream.
    /// 
    /// A unique marker is written to the serial port and then the port read buffer is consumed until the echo-back
    /// contains the unique marker, skipping any old or unexpected received data.
    /// It is then assumed that the next write can issue a valid command and expect a vaild response.
    fn resync(&mut self) -> Result<()> {
        debug!("resyncing serial connection...");

        use rand::distributions::{Alphanumeric, DistString};
        let marker = Alphanumeric.sample_string(&mut rand::thread_rng(), 8);
        let marker = format!("resync{}", marker);

        let cmd = format!("{}\r", marker);
        let reply = format!("{}\r\n#\r\nCommand Error.\r\n#", marker);

        println!("cmd: '{}', expected reply: '{}'", escape(&cmd), escape(&reply));

        self.port.write(cmd.as_bytes())?;
        self.read_until(reply.as_bytes())?;

        Ok(())
    }

    pub fn zone_enquiry(&mut self, id: ZoneId) -> Result<Vec<ZoneStatus>> {
        if let ZoneId::System = id {
            return id.to_amps().into_iter()
                .map(|amp| self.zone_enquiry(amp))
                .flatten_ok()
                .collect();
        }

        let (amp, zone, expected_responses) = match id {
            ZoneId::Zone { amp, zone } => (amp, zone, 1),
            ZoneId::Amp(amp) => (amp, 0, 6),
            ZoneId::System => unreachable!()
        };

        let cmd = format!("?{:}{:}", amp, zone);

        self.exec_command(cmd.as_bytes(), expected_responses)?
            .into_iter()
            .map(|resp| -> Result<ZoneStatus> {
            let values = resp[1..] // skip leading '>'
                .chunks_exact(2)
                .map(|c| -> Result<u8> {
                    let s = str::from_utf8(c).context("response string not valid UTF-8")?;

                    Ok(str::parse::<u8>(s).context("failed to parse u8")?)
                })
                .collect::<Result<Vec<_>>>()?;

            {
                use ZoneAttribute::*;

                Ok(ZoneStatus {
                    zone_id: ZoneId::try_from(values[0]).context("invalid zone id received from amp")?,
                    attributes: vec![
                        PublicAnnouncement(values[1] != 0),
                        Power(values[2] != 0),
                        Mute(values[3] != 0),
                        DoNotDisturb(values[4] != 0),
                        Volume(values[5]),
                        Treble(values[6]),
                        Bass(values[7]),
                        Balance(values[8]),
                        Source(values[9]),
                        KeypadConnected(values[10] != 0)
                    ] 
                })
            }
        }).collect()
    }

    pub fn set_zone_attribute(&mut self, id: ZoneId, attr: ZoneAttribute) -> Result<()> {
        if let ZoneId::System = id {
            return id.to_amps().into_iter()
                .map(|amp| self.set_zone_attribute(amp, attr))
                .collect();
        }

        attr.validate()?;

        let (attr, val) = {
            use ZoneAttribute::*;

            match attr {
                Power(v) => ("PR", v as u8),
                Mute(v) => ("MU", v as u8),
                DoNotDisturb(v) => ("DT", v as u8),
                Volume(v) => ("VO", v),
                Treble(v) => ("TR", v),
                Bass(v) => ("BS", v),
                Balance(v) => ("BL", v),
                Source(v) => ("CH", v),
                attr => bail!("{} cannot be changed", attr)
            }
        };


        let cmd = format!("<{}{}{:02}", id, attr, val);

        self.exec_command(cmd.as_bytes(), 0)?;

        Ok(())
    }
}